packages/arb-avm-cpp/avm_values/include/avm_values/code.hpp

```diff
@@ -135,25 +152,116 @@ struct CodeSnapshot {
+
+    virtual CodeSegmentSnapshot loadCodeSegment(uint64_t segment_num) const = 0;
+
+    virtual CodePoint loadCodePoint(const CodePointRef& ref) const = 0;
+
+    virtual CodePointStub addSegment() = 0;
+
+    virtual CodePointStub addOperation(const CodePointRef& ref,
+                                       Operation op) = 0;
+};
+
+template <typename T>
+class CodeBase {
+    const T* getThis() const { return static_cast<const T*>(this); }
+
+    T* getThis() { return static_cast<T*>(this); }
+
+   protected:
+    CodeSegmentSnapshot loadCodeSegmentImpl(uint64_t segment_num) const {
+        auto& segment = getThis()->getSegment(segment_num);
+        return {segment, segment->size(), segment->cached_hashes.size()};
+    }
+
+    CodePoint loadCodePointImpl(const CodePointRef& ref) const {
+        auto& segment = getThis()->getSegment(ref.segment);
+        return segment->loadCodePoint(ref.pc);
+    }
+
+    CodePointStub addSegmentImpl() {
+        uint64_t segment_num = getThis()->nextSegmentNum();
+        auto new_segment = std::make_shared<CodeSegment>(segment_num);
+        auto stub = new_segment->initialCodePointStub();
+        getThis()->storeSegment(std::move(new_segment));
+        return stub;
+    }
+
+    CodePointStub addOperationImpl(const CodePointRef& ref, Operation op) {
+        auto& segment = getThis()->getSegment(ref.segment);
+        auto initial_pc = segment->size() - 1;
+        if (ref.pc == initial_pc) {
+            if (segment.use_count() == 1) {
+                // No other code has a reference to this segment. That means we
+                // can modify the segment even if it forces a reallocation
+                return segment->addOperation(std::move(op));
+            }
+
+            if (segment->size() < segment->capacity()) {
+                // The segment has extra capacity so we can add to it even if
+                // other code has references to it
+                return segment->addOperation(std::move(op));
+            }
+
+            // Fall back to making a copy as there are other references and no
+            // space to add this operation
+        }
+        // This segment was already mutated elsewhere, therefore we must
+        // make a copy
+        uint64_t new_segment_num = getThis()->nextSegmentNum();
+        auto new_segment = segment->getSubset(new_segment_num, ref.pc);
+        auto stub = new_segment->addOperation(std::move(op));
+        getThis()->storeSegment(std::move(new_segment));
+        return stub;
+    }
+};
+
+class CoreCode : public CodeBase<CoreCode>, public Code {
+    friend CodeBase<CoreCode>;
+
     mutable std::mutex mutex;
     std::unordered_map<uint64_t, std::shared_ptr<CodeSegment>> segments;
     uint64_t next_segment_num;
 
+    const std::shared_ptr<CodeSegment>& getSegment(uint64_t segment_num) const {
+        return segments.at(segment_num);
+    }
+
+    uint64_t nextSegmentNum() { return next_segment_num++; }
```

Comment:
We might want to make this throw. I'm pretty sure it should never be called outside maybe tests? In practice we should be adding segments to a RunningCode and then committing them to the CoreCode. Directly adding them to CoreCode in non-test usage might mean a conflict with RunningCode which would be very bad I think.

Reply:
This is used in a lot of different tests so I think we hold off on this for now and fix in a separate PR later